File: Week01/1.MMRECT2/ikicha/mmrect2.cpp
#include <map>
#include <iostream>
#include <vector>
#include <algorithm>
//#include <fstream>
#include <set>
#include <cmath>
#define CONTAINS( x , y ) ((x).find(y) != (x).end())

using namespace std;

inline void solve(ifstream &fin);
inline void solve();
inline int min_flag(int x, int y);
int main() {
    int num_case;
    //ifstream fin("input.txt");
    //fin>> num_case;
    cin>> num_case;
    for(int c = 0; c < num_case; c++) { 
        //solve(fin);
        solve();
    }
    // fin.close();
    return 0;
}


//inline void solve(ifstream &fin) {
inline void solve() {
    map<int, set<int> > data;
    int num;
    //fin >> num;
    cin >> num;
    for(int i = 0; i < num; i++) {
        int temp1, temp2;    
        //fin >> temp1 >> temp2;
        cin >> temp1 >> temp2;
        data[temp1].insert(temp2);
    }
        
    map<int, set<int> >::iterator i;
    set<int>::iterator j, k;
    int max_result = 0, min_result = 0;
    for(i = data.begin(); i!=data.end();++i) {
        for(j = i->second.begin(); j!=i->second.end()--; ++j) {
            for(k = j; k!=i->second.end();++k) {
                if(*j != *k){
                   int size = abs(*j - *k);
                    for(int l = -1; l<2; l=l+2){
                        int flag = i->first + size * l;
                        if(CONTAINS(data, flag)) {
                            if(CONTAINS(data[flag], *j) && CONTAINS(data[flag], *k)) {
                                max_result = max(max_result, size);
                                min_result = min_flag(min_result, size);
                            }
                        }
                    }
                }
            }

        }
        
    }
    cout << min_result << " " << max_result << endl;
}

inline int min_flag(int x, int y) {
    return x==0? y: min(x,y);
}

File: Week01/1.MMRECT2/ikicha/mmrect2.py
def solve():#f):
    n = input()
#    n = eval(f.readline())
    data = {}
    for i in range(0, n):
        temp = map(eval, raw_input().split(" "))
#        temp = map(eval, f.readline().split())
        if not temp[0] in data.keys():
            data[temp[0]] = []

        data[temp[0]].append(temp[1])

    min_result = -1
    max_result = 0
    for points in data.keys():
        data_point = data[points]
        for left in range(0, len(data_point)):
            for right in range(left + 1, len(data_point)):
                size = abs(data_point[left] - data_point[right])
                if points + size in data:
                    if data_point[left] in data[points + size] and data_point[right] in data[points + size]:
                        max_result = max(max_result, size)
                        min_result = min_flag(min_result, size)

                if points - size in data:
                    if data_point[left] in data[points - size] and data_point[right] in data[points - size]:
                        max_result = max(max_result, size)
                        min_result = min_flag(min_result, size)
        
    print str(min_result) + " " + str(max_result)

def min_flag(x, y):
    if x==-1:
        return y
    else:
        return min(x,y)
               



#f = open("input.txt")
num_case = input()
#num_case = eval(f.readline())

for c in range(0, num_case):
    #solve(f)
    solve()
#f.close()
